Adds FlagserPersistence

gtauzin opened the pull request:
#### What does this implement/fix? Explain your changes.
- Adds `FlagserPersistence` class and `pyflagser` as a library requirement.
- Fixes docstrings for some classes in `simplicial.py`.
- Fixes a bug leading to features with negative lifetime in persistent homology transformers when `infinity_values` is set too low.
- Clean up code in `_utils.py`.

Commit:
[WIP] Add FlagPersistence

Signed-off-by: Guillaume Tauzin <guillaume.tauzin@epfl.ch>

Commit:
Merge with upstream master

Signed-off-by: Guillaume Tauzin <guillaume.tauzin@epfl.ch>

Commit:
Update FlagPersistence

Signed-off-by: Guillaume Tauzin <guillaume.tauzin@epfl.ch>

Commit:
[WIP] Add tests

Signed-off-by: Guillaume Tauzin <guillaume.tauzin@epfl.ch>

Commit:
Merge with upstream master

Signed-off-by: Guillaume Tauzin <guillaume.tauzin@epfl.ch>

Commit:
Rename homology files

Signed-off-by: Guillaume Tauzin <guillaume.tauzin@epfl.ch>

Commit:
Add FlagPersistence to doc index

Signed-off-by: Guillaume Tauzin <guillaume.tauzin@epfl.ch>

wreise commented on gtda/homology/simplicial.py:
```diff
@@ -612,3 +613,202 @@ def transform(self, X, y=None):
+class FlagPersistence(BaseEstimator, TransformerMixin):
+    """`Persistence diagrams <https://giotto.ai/theory>`_ resulting from
+    `filtrations of directed or undirecte flag complexes
+    <https://giotto.ai/theory>`_.
```
Typo: undirected

wreise commented on gtda/homology/simplicial.py:
```diff
@@ -612,3 +613,202 @@ def transform(self, X, y=None):
+    See also
+    --------
+    ConsistentRescaling
+
```
Why?  What about VR and all?

gtauzin replied:
Agreed.

I should add that those entries need to be updated throughout the package. There are no examples and references are missing too. Also we need to show graphic examples once the plotting API is ready.

wreise commented on gtda/homology/simplicial.py:
```diff
@@ -612,3 +613,202 @@ def transform(self, X, y=None):
+    -----
+    `Flagser <https://github.com/luetge/flagser>`_ is used as a C++ backend
+    for computing Flag complexes persistent homology. Python bindings rely on
+    the `pyflagser <https://github.com/giotto-ai/pyflagser>`_ package.
```
What about putting forward the binding, *which  uses `Flagser` package*? No strong feelings for this from my side though.
For example, something along the lines of:
This is an adapter for `pyflagser <https://github.com/giotto-ai/pyflagser>`_, the Python bindings for `Flagser <https://github.com/luetge/flagser>`_ , the C++ backend for computing Flag complexes persistent homology.

gtauzin replied:
I have updated the description.

wreise commented on gtda/homology/simplicial.py:
```diff
@@ -612,3 +613,202 @@ def transform(self, X, y=None):
+    _hyperparameters = {'directed': [bool, [True, False]],
+                        'max_entries': [int, None],
+                        'max_edge_length': [numbers.Number, None],
+                        'infinity_values_': [numbers.Number],
```
Don't we need a  `None` for the constraints?

wreise commented on gtda/homology/simplicial.py:
```diff
@@ -612,3 +613,202 @@ def transform(self, X, y=None):
+                        approximation=self.max_entries)['dgms']
+
+        if 0 in self._homology_dimensions:
+            Xdgms[0] = Xdgms[0][:-1, :]  # Remove final death at np.inf
```
A negligible detail: maybe `Xdgms[0].slice(0,-1)` ?

gtauzin replied:
This is everywhere. So same answer as above. Documentation and code needs to be consistently updated throughout the package.

gtauzin replied:
But I am also not sure why it is better?

Commit:
Add hyperparameter range

Commit:
Fix typo

Commit:
Fix doc

ulupo committed:
Rename class to FlagserPersistence, add See Also entries in simplicial (#15)

Commit:
Merge branch 'master' into ulupo_flag

Commit:
Linting in ripser_interface

Commit:
Fixes in FlagserPersistence

- Several fixes in docstrings
- Adapt to new hyperparameter validation
- add check_is_fitted step in `transform`

Commit:
Attempt to fix FlagserPersistence tests

Test still fails as the output of fit_transform is a trivial list (not an ndarray as documented in pyflagser, and not equal to the output of VietorisRipsPersistence when direceted is False). Possibly pointing to a serious bug in pyflagser @gtauzin ?

gtauzin committed:
Add filtration to docstring

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

gtauzin committed:
Merge branch 'flag' of github.com:gtauzin/giotto-tda into flag

ulupo committed:
Bump pyflagser version to 0.2.1 in requirements.txt

Commit:
Merge branch 'master'

Conflicts:
	doc/index.rst
	gtda/externals/python/ripser_interface.py
	gtda/homology/simplicial.py

ulupo committed:
Add pyflagser to GH pages installation instructions

ulupo committed:
Merge branch 'master' into flag

ulupo committed:
Fix tests syntax

ulupo committed:
Merge branch 'master' into flag

ulupo committed:
Fix variable name

ulupo committed:
Implement input checks in FlagserPersistence via check_point_clouds

wreise commented on gtda/homology/simplicial.py:
```diff
@@ -720,3 +721,230 @@ def plot(Xt, sample=0, homology_dimensions=None):
+        Upper bound on the maximum value of the filtration parameter. Points
+        whose distance is greater than this value will never be connected by
+        an edge, and topological features at scales larger than this value
+        will not be detected.
```
I would interpret the description as: we stop the filtration at `max_edge_length`. However, i am sure that i misunderstand, because I don't think this is what happens.
Also, do we test this behavior?

ulupo replied:
This should be what happens. It should be the same behaviour as `VietorisRipsPersistence`.

ulupo replied:
@wreise could you explain the origin of the doubt as to whether this is really what happens?

wreise replied:
Ok, nothing. I forgot that we apply post-processing where we replace `np.inf`, which by default is `max_edge_length`

ulupo replied:
@wreise actually I now think you are right, having had some time to look into it better. It *looks* like this is not achieving quite the same as in `VietorisRipsPersistence`. In particular, before postprocessing is applied I do not think one will have infinite bars at all, and in particular no infinities which can be replaced with `infinity_values`.

wreise replied:
🤦‍♂ i had checked only for `ViertorisRipsPersistence` in the end, where it is ok.
Yes, we get unexpected results when, for example, `max_edge_length` and `infinity_values=np.inf`are both passed.
To be more specific, in
```
Xs = np.array([[[0, 1, 10, 1],
                [1, 0, 1, 10],
                [10, 1, 0, 1],
                [1, 10, 1, 0]]]).astype(float)

FlagserPersistence(max_edge_length=3, infinity_values=np.inf).fit_transform(Xs)
```
generators in H_1 die at 3, instead of `np.inf`.

ulupo committed:
Fix variable name in transform, add PlotterMixin inheritance

wreise commented on gtda/homology/simplicial.py:
```diff
@@ -720,3 +721,230 @@ def plot(Xt, sample=0, homology_dimensions=None):
+    def _flagser_diagram(self, X):
+        X[X >= self.max_edge_length] = self.max_edge_length
+
+        Xdgms = flagser(X[X[:, 0] != np.inf],
```
What is `X[:, 0] != np.inf` and why do we exclude those rows?

ulupo replied:
There is a historical reason for this that maybe @gtauzin remembers but I can't. I think it had to do with the implementation of some of the classes in the images submodule. Perhaps `ImageToPointCloud` which used to return arrays with some infinities in them?

ulupo replied:
(It is the same almost everywhere else in `gtda.homology`)

wreise replied:
It seems quite dangerous, when (and only when!) `metric='precomputed'`, in both `VietorisRipsPersistence` and `FlagserPersistence`.
Take the following example:
```
Xs = np.array([[[0, 1, np.inf, 1],
              [1, 0, 1, 10],
              [np.inf, 1, 0, 1],
              [1, 10, 1, 0]]]).astype(float)
```
Then, `FlagserPersistence().fit_transform(Xs)` works if we remove `X[:, 0] != np.inf`. Otherwise, it removes one row from the distance matrix and you get something not square  - strangely, the shape of the input is not validated, and the kernel crashes.

VietorisRips, in its current state, behaves as Flagser. But, it does not handle`np.inf`, so we may want to find another solution

ulupo replied:
Apparently (having checked with @gtauzin) whatever reason there ever was is no longer relevant. Good, because it really shouldn't be there. I'll remove it. Though I didn't quite understand your latest message.

wreise replied:
I am just pointing out that if you remove the 'input masking' in ripser as well, VietorisRipsPersistence will (still) fail on the input above.
Not saying this is unexpected behavior: maybe `np.inf` in the distance matrix is a valid reason for it to fail.

ulupo replied:
@wreise am I right to believe that this discussion can be resolved following #386 ?

ulupo committed:
Fix duplicate test names

ulupo committed:
Remove obsolete input maskings in calls to ripser and flagser

Commit:
Merge branch 'master' into flag

wreise committed:
Change the parameter name in the validation in the Flagser transformer

wreise committed:
Change the matrix used in a failed test to a square one, so as the test fails for a valid reason

wreise committed:
Merge branch 'master' into g_flag

ulupo committed:
Change input in test_fp_not_fitted just for consistency

gtauzin committed:
Merge with remote

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

gtauzin committed:
Merge branch 'master' of github.com:giotto-ai/giotto-tda into flag

gtauzin committed:
Merge with remote

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

gtauzin committed:
Use pyflagser v0.3.0

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

gtauzin committed:
Update FlagserPersistence doc and code based on pyflagser v0.3.0 improvements

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

gtauzin committed:
Fix handling of max_edge_weight

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

ulupo committed:
Bump pyflagser requirement to 0.3.1

gtauzin committed:
Set pyflagser version to 0.3.1

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

gtauzin committed:
Merge with remote

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

ulupo commented on README.rst:
```diff
@@ -67,6 +67,7 @@ The latest stable version of ``giotto-tda`` requires:
 - SciPy (>= 0.17.0)
 - joblib (>= 0.13)
 - scikit-learn (>= 0.22.0)
+- pyflagser (>- 0.3.1)
```
Typo with ">-"?

gtauzin replied:
Thanks!

gtauzin committed:
Fix typo

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

gtauzin committed:
Fix docstrings typo

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

gtauzin committed:
Reattach hera's head

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

ulupo commented on gtda/homology/simplicial.py:
```diff
@@ -86,8 +87,8 @@ class VietorisRipsPersistence(BaseEstimator, TransformerMixin, PlotterMixin):
     --------
-    SparseRipsPersistence, EuclideanCechPersistence, CubicalPersistence,
-    ConsistentRescaling
+    FlagserPersistence, SparseRipsPersistence, EuclideanCechPersistence, \
```
Would you also think that simplicial persistence transformers should not be referred to in the "See also" section of `CubicalPersistence`?

gtauzin replied:
Yes, I think this was the case when we only had `CubicalPersistence` and `VietorisRipsPersistence`.

I agree, `CubicalPersistence` should not be in the "See also" section of simplicial transformers and simplicial transformers and vice-versa.

ulupo committed:
Fix refs

gtauzin committed:
Update See also for Cubical

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

gtauzin committed:
Merge branch 'flag' of github.com:gtauzin/giotto-tda into flag

ulupo commented on gtda/homology/simplicial.py:
```diff
@@ -721,3 +722,272 @@ def plot(Xt, sample=0, homology_dimensions=None):
+        detected.
+
+    directed : bool, optional, default: ``True``
+        If ``True``, :meth:`transform` computes the persistence diagrams of
```
I'd like to suggest a rewording of this section as follows:
```
        If ``True``, :meth:`transform` computes the persistence diagrams of
        the filtered directed flag complexes arising from the input collection
        of weighted directed graphs. If ``False``, :meth:`transform` computes
        the persistence diagrams of the filtered undirected flag complexes
        obtained by regarding all input weighted graphs as undirected. When
        ``False``, it is sufficient to pass a collection of upper-triangular
        matrices as input to :meth:`transform` because, if two directed edges
        corresponding to the same undirected edge are assigned different
        weights, only the one in the upper triangular part of the adjacency
        matrix is considered.
```

gtauzin replied:
The logic of the last sentence is not clear to me:

>        When ``False``, it is sufficient to pass a collection of upper-triangular
>         matrices as input to :meth:`transform` because, if two directed edges
>         corresponding to the same undirected edge are assigned different
>         weights, only the one in the upper triangular part of the adjacency
>         matrix is considered.

It is sufficient because the weight on only one of the oriented edge is enough to assign a weight to the unoriented edge. The last part of the sentence deals with the edge case where two oriented edges corresponding to the same unoriented edge have different weights.

ulupo replied:
I still see it slightly differently:
- If the input is dense, there are always two explicit weights per undirected edge;
- If the input is sparse, there are still two weights per undirected edge because we say in the sparse case that absence of entry (i, j) is actually meaningful (i.e. it means infinite directed edge i->j).

ulupo replied:
If we referred to the docstrings for transform here then my second bullet point would be more compelling I think.

gtauzin replied:
I think the reality is more complicated:

- If the input is dense and not symmetric but some values of the upper-trianguar part are larger than `max_edge_weight`, the value on the lower-trinagular part will be considered. This is also true in the sparse case actually.
- If the inpur is sparse, and there is no entry on an edge in the upper-triangular part, the value of the unoriented edge will be using the weight in the lower-triangular part.

ulupo replied:
Thanks for the feedback. What's the reason for the behaviour in your first bullet point? It seems a bit confusing to me.

gtauzin replied:
If the weight is higher than `max_edge_weight`, it is absent from the filtration and therefore not passed to C++ `flagser`.

ulupo replied:
Sure, but why is the lower weight picked instead? In my opinion, the first thing we should do in the dense case is check for disagreements between off-diagonal entries, and resolve them in favour of the upper diagonal. Then the issue you describe would not arise.

ulupo replied:
But this is what we have for now so we are stuck with documenting this weird (from the user perspective) behaviour. I'll try to come up with an alternative!

ulupo replied:
```
        If ``True``, :meth:`transform` computes the persistence diagrams of
        the filtered directed flag complexes arising from the input collection
        of weighted directed graphs. If ``False``, :meth:`transform` computes
        the persistence diagrams of the filtered undirected flag complexes
        obtained by regarding all input weighted graphs as undirected, and:

        - if `max_edge_weight` is ``numpy.inf``, it is sufficient to pass a
          collection of (dense or sparse) upper-triangular matrices;
        - if `max_edge_weight` is finite, it is recommended to pass either a
          collection of symmetric dense matrices, or a collection of sparse
          upper-triangular matrices;
        - if two directed edges corresponding to the same undirected edge are
          explicitly assigned different weights and neither exceeds
          `max_edge_weight`, only the one in the upper triangular part of the
          adjacency matrix is considered.
```

What do you think? If you agree, it might be a good idea to reference to this again in the docstrings for `fit` and `transform`.

ulupo replied:
I feel that with something like the above we would be guiding the user a little more directly.

gtauzin replied:
As I read this, it definitely sounds too complex to be fully explicitly documented. I would suggest not explaining all those rare edge cases but rather explain the user what type of adjacency matrix has to be passed and why. 

```
        If ``True``, :meth:`transform` computes the persistence diagrams of
        the filtered directed flag complexes arising from the input collection
        of weighted directed graphs. If ``False``, :meth:`transform` computes
        the persistence diagrams of the filtered undirected flag complexes
        obtained by regarding all input weighted graphs as undirected. In
        that case, if two directed edges corresponding to the same 
        undirected edge are explicitly assigned different weights and neither 
        exceeds `max_edge_weight`, only the one in the upper triangular part 
        of the adjacency matrix is considered. Therefore, it is recommended to 
        pass either a collection of symmetric or upper-triangular matrices.
```

An alternative would also be to force a simpler behavior by systematically enforcing that the lower triangular part of the matrix is equal to the upper triangular part.

ulupo replied:
Thanks for the feedback. I agree about being prescriptive rather than comprehensively descriptive, but in that case I would take a different route and avoid altogether to discuss what happens when different weights are assigned. I'd simply tell the user which behaviour is safe. Simply removing the final point in the suggestion I had above would do the job:
```
        If ``True``, :meth:`transform` computes the persistence diagrams of
        the filtered directed flag complexes arising from the input collection
        of weighted directed graphs. If ``False``, :meth:`transform` computes
        the persistence diagrams of the filtered undirected flag complexes
        obtained by regarding all input weighted graphs as undirected, and:

        - if `max_edge_weight` is ``numpy.inf``, it is sufficient to pass a
          collection of (dense or sparse) upper-triangular matrices;
        - if `max_edge_weight` is finite, it is recommended to pass either a
          collection of symmetric dense matrices, or a collection of sparse
          upper-triangular matrices.
```

Sorry to be a pain! I do think we have come a long way since the first versions though :)

gtauzin replied:
Agreed, that works.

ulupo commented on gtda/homology/simplicial.py:
```diff
@@ -721,3 +722,272 @@ def plot(Xt, sample=0, homology_dimensions=None):
+        matrix is considered.
+
+    filtration : string, optional, default: ``'max'``
+        Algorithm determining the filtration. Warning: if an edge filtration is
```
What does this warning really mean to the user? Presumably the "possible values" below all make "consistent filtrations", and if the user cannot pass custom algorithms then why should we tell them this?

gtauzin replied:
I believe this means that: "It is assumed that a vertex weight cannot be larger than the weight of the edges it forms."

ulupo replied:
But then this doesn't have anything to do with the filtration algorithm and all to do with the input right?

gtauzin replied:
Wait, no, I was wrong. I do believe that this was for the custom filtration that C++ `flagser` allows.

My comment above does apply to the input indeed.

ulupo replied:
Ok, so how about adding your sentence above to the docs for `X` in `fit` and `transform`? And removing this warning entirely from here?

gtauzin replied:
Done, thanks!

ulupo commented on gtda/homology/simplicial.py:
```diff
@@ -721,3 +722,272 @@ def plot(Xt, sample=0, homology_dimensions=None):
+        'coeff': {'type': int, 'in': Interval(2, np.inf, closed='left')},
+        'max_edge_weight': {'type': Real},
+        'infinity_values': {'type': (Real, type(None))},
+        'max_entries': {'type': int}
```
Add list of algorithms? (See comment above.)

gtauzin replied:
The validaty of the filtration algorithm is already checked by `pyflagser`.

ulupo replied:
Thanks!

ulupo committed:
Fix max_edge_length docs in other simplicial transformers

gtauzin committed:
Fix docstring typo

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

ulupo commented on gtda/homology/simplicial.py:
```diff
@@ -721,3 +722,272 @@ def plot(Xt, sample=0, homology_dimensions=None):
+    max_edge_weight : float, optional, default: ``numpy.inf``
+        Maximum edge weight to be considered in the filtration. All edge
+        weights greater than that value will be considered as
+        infinitely-valued, i.e., absent from the filtration and topological
```
I think "infinitely-valued" can be a little confusing here because in fact if you play with `infinity_values` you can change this. Do you agree?

gtauzin replied:
I will write some extra test and compare the meaning of `infinty_values` in VR and Flagser as per your main comment and come back to you.

ulupo replied:
Thanks!

gtauzin replied:
Agreed, I simply removed "infintely-valued".

ulupo replied:
Thanks!

ulupo committed:
Remove extra whitespace

ulupo reviewed (changes requested):
Do you think that @wreise and my earlier concerns that the behaviour of `max_edge_weight` and `infinity_values` were not consistent with what happens in `VietorisRipsPersistence` are finally resolved? I would think so for `infinity_values` since the code for transform looks identical in the two classes, but since we defer to `pyflagser` for treatment of `max_edge_weight`, maybe it is worth testing this/cross checking the `pyflagser` and `VietorisRipsPersistence` code?

ulupo commented on gtda/homology/cubical.py:
```diff
@@ -68,7 +68,9 @@ class CubicalPersistence(BaseEstimator, TransformerMixin, PlotterMixin):
     See also
     --------
-    VietorisRipsPersistence, SparseRipsPersistence, EuclideanCechPersistence
+    images.HeightFiltration, images.RadialFiltration, \
```
Thanks!

gtauzin committed:
Improve filtration docstring

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

gtauzin committed:
Merge branch 'flag' of github.com:gtauzin/giotto-tda into flag

gtauzin committed:
Improve input description

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

gtauzin committed:
Merge branch 'master' of github.com:giotto-ai/giotto-tda into flag

gtauzin committed:
Fix max_edge_weight docstring inconsistency

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

gtauzin committed:
Improve directed docstring

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

ulupo committed:
Grammar fix

gtauzin committed:
Add tests to check max_edge and infinity_values

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

gtauzin committed:
Merge branch 'flag' of github.com:gtauzin/giotto-tda into flag

gtauzin committed:
Fix low infinity value bug and put dimension padding in _utils

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

gtauzin committed:
Add test for low infinity values

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

gtauzin committed:
Propagate the change of dim padding to cubical

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

gtauzin committed:
Fix padding bug

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

ulupo commented on gtda/homology/_utils.py:
```diff
@@ -16,12 +16,29 @@ def _pad_diagram(Xd, homology_dimensions, max_n_points, min_values):
 
 def _postprocess_diagrams(Xt, homology_dimensions, infinity_values, n_jobs):
+    # Replacing np.inf with infinity_values
+    Xt = [{dim: np.nan_to_num(Xt[i][dim], posinf=infinity_values)
```
Simpler version:
```
    Xt = [{dim: np.nan_to_num(diagram[dim], posinf=infinity_values)
          for dim in homology_dimensions}
          for diagram in Xt]
```

ulupo commented on gtda/homology/_utils.py:
```diff
@@ -16,12 +16,29 @@ def _pad_diagram(Xd, homology_dimensions, max_n_points, min_values):
+          for i in range(len(Xt))]
+
+    # Removing points whose birth is higher than their death
+    Xt = [{dim: Xt[i][dim][Xt[i][dim][:, 0] < Xt[i][dim][:, 1]]
```
Simpler-looking version:
```
    Xt = [{dim: subdiagram[subdiagram[:, 0] < subdiagram[:, 1]]
          for dim, subdiagram in diagram.items()}
          for diagram in Xt]
```

ulupo commented on gtda/homology/_utils.py:
```diff
@@ -16,12 +16,29 @@ def _pad_diagram(Xd, homology_dimensions, max_n_points, min_values):
 
 
 def _postprocess_diagrams(Xt, homology_dimensions, infinity_values, n_jobs):
+    # Replacing np.inf with infinity_values
```
Am I right to think this is the first time that each diagram is turned into a dictionary? I believe before this `Xt` is a list of outputs of `ripser` which are lists of arrays. Correct? If so, maybe we can add this to the comment?

ulupo commented on gtda/homology/_utils.py:
```diff
@@ -16,12 +16,29 @@ def _pad_diagram(Xd, homology_dimensions, max_n_points, min_values):
+          for dim in homology_dimensions}
+          for i in range(len(Xt))]
+
     max_n_points = {dim: max(1, np.max([Xt[i][dim].shape[0]
```
Simpler-looking version:
```
    max_n_points = {dim: np.max([len(diagram[dim]) for diagram in Xt] + [1])
                    for dim in homology_dimensions}
```

ulupo replied:
Also some similar simplifications for lines 28-32:
```
    min_values = {dim: min([np.min(diagram[dim][:, 0]) if diagram[dim].size else
                            np.inf for diagram in Xt])
                  for dim in homology_dimensions}
    min_values = {dim: min_value if min_value != np.inf else 0
                  for dim, min_value in min_values.items()}
```

ulupo commented on gtda/homology/_utils.py:
```diff
@@ -34,4 +51,4 @@ def _postprocess_diagrams(Xt, homology_dimensions, infinity_values, n_jobs):
         Xt[i], homology_dimensions, max_n_points, min_values)
```
Similarly to the above, can do:
```
    Xt = Parallel(n_jobs=n_jobs)(delayed(_pad_diagram)(
        diagram, homology_dimensions, max_n_points, min_values)
                                      for diagram in Xt)
```

ulupo commented on gtda/homology/_utils.py:
```diff
@@ -16,12 +16,29 @@ def _pad_diagram(Xd, homology_dimensions, max_n_points, min_values):
+                                            dtype=Xd[dim].dtype)])
+          for dim in homology_dimensions]
+
+    Xd = np.vstack([Xd[d] for d in range(len(homology_dimensions))])
```
This looks equivalent to simply doing:
```
   Xd = np.vstack(Xd)
```
because `Xd` is defined as a list above.

ulupo commented on gtda/homology/_utils.py:
```diff
@@ -16,12 +16,29 @@ def _pad_diagram(Xd, homology_dimensions, max_n_points, min_values):
+                [min_values[dim], min_values[dim]]
+
+    # Add dimension as the third elements of each (b, d) tuple
+    Xd = [np.hstack([Xd[dim], dim * np.ones((Xd[dim].shape[0], 1),
```
Simpler-looking alternative:
```
    Xd = [np.insert(Xd[dim], 2, dim, axis=1)
          for dim in homology_dimensions]
```

gtauzin replied:
Nice!

ulupo reviewed (commented):
Just a few style suggestions.

gtauzin committed:
Simplifyl code in utils

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

ulupo reviewed (changes requested):
@gtauzin It looks like you executed the voids on the plane example and committed the output cells. It looks to me like this was done accidentally and any changes to this notebook should be reverted. Am I mistaken?

gtauzin committed:
Revert voids notebook commit

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

ulupo commented on examples/voids_on_the_plane.ipynb:
```diff
@@ -32,7 +32,7 @@
     "import itertools\n",
     "\n",
-    "import matplotlib.pyplot as plt  # Not a requirement of giotto-tda, but is needed here\n",
+    "import matplotlib.pyplot as plt\n",
```
Could you reinstate the comment on matplotlib?

gtauzin committed:
Merge branch 'master' of github.com:giotto-ai/giotto-tda into flag

gtauzin committed:
Readd comment on matplotlib

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

gtauzin committed:
Bump pyflagser version to 0.4.0

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

gtauzin committed:
Extend test to empty subdiagrams

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

ulupo committed:
Fix comparison with literal

ulupo reviewed (approved):
I've updated the description of this PR to include some of the extra changes that were made in the process.  So glad this is finally happening! Thanks for all the work @gtauzin!
